File: app/src/main/java/com/example/tanawatk/mvvmandroid/service/repo/local/NewsLocalDataSource.java
package com.example.tanawatk.mvvmandroid.service.repo.local;

import com.example.tanawatk.mvvmandroid.common.RequestFailureError;
import com.example.tanawatk.mvvmandroid.common.ResponseHandler;
import com.example.tanawatk.mvvmandroid.service.model.ResponseModel;
import com.example.tanawatk.mvvmandroid.service.repo.NewsDataSource;

import javax.inject.Inject;

public class NewsLocalDataSource implements NewsDataSource {

    @Inject
    public NewsLocalDataSource() {
    }

    @Override
    public void getNews(int id, ResponseHandler<ResponseModel> responseHandler) {
        responseHandler.onRequestFailure(new RequestFailureError(0,"Now we don't have local database"));
    }

    @Override
    public void loadNews(ResponseHandler<ResponseModel> responseHandler) {

    }

    @Override
    public void loadDatabase(ResponseHandler<ResponseModel> responseHandler) {

    }

    @Override
    public void saveDatabase(ResponseModel model) {

    }

    @Override
    public void deleteDatabase() {

    }
}

File: app/src/main/java/com/example/tanawatk/mvvmandroid/service/repo/local/NewsLocalDataSource.kt
package com.example.tanawatk.mvvmandroid.service.repo.local

import com.example.tanawatk.mvvmandroid.common.Result
import com.example.tanawatk.mvvmandroid.service.db.NewsDao
import com.example.tanawatk.mvvmandroid.service.db.NewsEntity
import com.example.tanawatk.mvvmandroid.service.model.News
import com.example.tanawatk.mvvmandroid.service.model.ResponseModel
import com.example.tanawatk.mvvmandroid.service.repo.LocalDataSource
import javax.inject.Inject

class NewsLocalDataSource @Inject constructor(
    private val newsDao: NewsDao
) : LocalDataSource {

    override suspend fun loadFromCache(): Result<ResponseModel> {
        val entities = newsDao.getAll()
        return if (entities.isEmpty()) {
            Result.Error(0, "Cache empty")
        } else {
            Result.Success(ResponseModel(null, null, 0, null, entities.map { it.toNews() }))
        }
    }

    override suspend fun replaceCache(model: ResponseModel) {
        val entities = model.articles?.map { it.toEntity() } ?: return
        newsDao.replaceAll(entities)  // single @Transaction — no gap between clear and insert
    }

    private fun NewsEntity.toNews() = News(
        url = url, title = title, date = date,
        authorName = authorName, authorUrl = null, forumUrl = null,
        imageUrl = imageUrl, comments = comments, intro = intro
    )

    private fun News.toEntity() = NewsEntity(
        url = url, title = title, date = date,
        authorName = authorName, imageUrl = imageUrl,
        intro = intro, comments = comments
    )
}